File: 16wrap/Views/StreamView/StreamView.m
//
//  TableView.m
//  ScrollAnimation
//
//  Created by Ravenpod on 04.02.13.
//  Copyright (c) 2013 Ravenpod. All rights reserved.
//

#import "StreamView.h"
#import "StreamLayout.h"
#import "StreamMetrics.h"

static NSString *StreamViewCommonLocksChanged = @"StreamViewCommonLocksChanged";

@interface StreamView ()

@property (nonatomic) NSInteger numberOfSections;

@property (strong, nonatomic) NSMutableSet *items;

@property (nonatomic) BOOL reloadAfterUnlock;

@property (nonatomic) NSUInteger locks;

@end

@implementation StreamView

@synthesize layout  = _layout;

@dynamic delegate;

- (void)dealloc {
	[self removeObserver:self forKeyPath:@"contentOffset"];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:StreamViewCommonLocksChanged object:nil];
}

- (void)setup {
    _items = [NSMutableSet set];
	UITapGestureRecognizer *tapRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tap:)];
	[self addGestureRecognizer:tapRecognizer];
	[self addObserver:self forKeyPath:@"contentOffset" options:NSKeyValueObservingOptionNew context:NULL];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(locksChanged) name:StreamViewCommonLocksChanged object:nil];
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
	[self updateVisibility];
}

- (id)initWithFrame:(CGRect)frame {
	self = [super initWithFrame:frame];
	if (self) {
		[self setup];
	}
	return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder {
    self = [super initWithCoder:coder];
    if (self) {
        [self setup];
    }
    return self;
}

- (StreamLayout *)layout {
    if (!_layout) {
        self.layout = [[StreamLayout alloc] init];
    }
    return _layout;
}

- (void)setLayout:(StreamLayout *)layout {
	_layout = layout;
	layout.streamView = self;
}

- (BOOL)horizontal {
    return self.layout.horizontal;
}

- (void)setHorizontal:(BOOL)horizontal {
    self.layout.horizontal = horizontal;
}

- (void)clear {
    for (StreamItem *item in _items) {
        [item.view removeFromSuperview];
    }
    [_items removeAllObjects];
}

static NSUInteger commonLocks = 0;

+ (void)lock {
    commonLocks = MAX(0, commonLocks + 1);
}

+ (void)unlock {
    if (commonLocks > 0) {
        commonLocks = commonLocks - 1;
        [[NSNotificationCenter defaultCenter] postNotificationName:StreamViewCommonLocksChanged object:nil];
    }
}

- (void)locksChanged {
    if (self.locks == 0 && commonLocks == 0 && self.reloadAfterUnlock) {
        self.reloadAfterUnlock = NO;
        [self reload];
    }
}

- (void)lock {
    self.locks = MAX(0, self.locks + 1);
}

- (void)unlock {
    if (self.locks > 0) {
        self.locks = self.locks - 1;
        [self locksChanged];
    }
}

- (void)reload {
    
    if (self.locks > 0 && commonLocks > 0) {
        self.reloadAfterUnlock = YES;
        return;
    }
    
    [self clear];
    
    StreamLayout *layout = self.layout;
    
    [layout prepare];
    
    id <StreamViewDelegate> delegate = self.delegate;
    
    if ([delegate respondsToSelector:@selector(streamViewNumberOfSections:)]) {
        self.numberOfSections = [delegate streamViewNumberOfSections:self];
    } else {
        self.numberOfSections = 1;
    }
    
    if ([delegate respondsToSelector:@selector(streamViewHeaderMetrics:)]) {
        NSArray *headers = [delegate streamViewHeaderMetrics:self];
        [self layoutMetrics:headers layout:layout at:nil];
    }
    
    NSUInteger count = _items.count;
    
    for (NSUInteger section = 0; section < self.numberOfSections; ++section) {
        
        StreamIndex *sectionIndex = [StreamIndex index:section];
        
        if ([delegate respondsToSelector:@selector(streamView:sectionHeaderMetricsInSection:)]) {
            NSArray *headers = [delegate streamView:self sectionHeaderMetricsInSection:section];
            [self layoutMetrics:headers layout:layout at:sectionIndex];
        }
        
        NSInteger numberOfItems = [delegate streamView:self numberOfItemsInSection:section];
        
        for (NSUInteger i = 0; i < numberOfItems; ++i) {
            StreamIndex *index = [[sectionIndex copy] add:i];
            NSArray *metrics = [delegate streamView:self metricsAt:index];
            [self layoutMetrics:metrics layout:layout at:index];
        }
        
        if ([delegate respondsToSelector:@selector(streamView:sectionFooterMetricsInSection:)]) {
            NSArray *footers = [delegate streamView:self sectionFooterMetricsInSection:section];
            [self layoutMetrics:footers layout:layout at:sectionIndex];
        }
        
        [layout prepareForNextSection];
    }
    
    BOOL empty = _items.count == count;
    
    if ([delegate respondsToSelector:@selector(streamViewFooterMetrics:)]) {
        NSArray *footers = [delegate streamViewFooterMetrics:self];
        [self layoutMetrics:footers layout:layout at:nil];
    }
    
    if (empty) {
        if ([delegate respondsToSelector:@selector(streamViewPlaceholderMetrics:)]) {
            StreamMetrics *placeholder = [delegate streamViewPlaceholderMetrics:self];
            if (placeholder) {
                CGSize size = self.frame.size;
                UIEdgeInsets insets = self.contentInset;
                if (self.horizontal) {
                    placeholder.size = size.width - insets.left - insets.right - layout.contentSize.width;
                } else {
                    placeholder.size = size.height - insets.top - insets.bottom - layout.contentSize.height;
                }
                [self layout:layout metrics:placeholder at:nil];
            }
        }
    }
    
    [layout finalize];
    
    self.contentSize = layout.contentSize;
    
    [self updateVisibility];
}

- (void)layoutMetrics:(NSArray*)metrics layout:(StreamLayout*)layout at:(StreamIndex*)index {
    for (StreamMetrics *m in metrics) {
        [self layout:layout metrics:m at:index];
    }
}

- (void)layout:(StreamLayout*)layout metrics:(StreamMetrics*)metrics at:(StreamIndex*)index {
    if (![metrics hiddenAt:index]) {
        StreamItem *item = [[StreamItem alloc] init];
        item.index = index;
        item.metrics = metrics;
        [layout layout:item];
        if (!CGSizeEqualToSize(item.frame.size, CGSizeZero)) {
            [_items addObject:item];
        }
    }
}

- (id)viewForItem:(StreamItem *)item {
    
    StreamReusableView *view = [item.metrics loadView];
    
    if (view) {
        view.frame = item.frame;
        item.view = view;
        id entry = nil;
        if ([self.delegate respondsToSelector:@selector(streamView:entryAt:)]) {
            entry = [self.delegate streamView:self entryAt:item.index];
        }
        if (item.metrics.prepareAppearingBlock) item.metrics.prepareAppearingBlock(item, entry);
        view.entry = entry;
        if (item.metrics.finalizeAppearingBlock) item.metrics.finalizeAppearingBlock(item, entry);
        view.frame = item.frame;
    }
    
    return view;
}

- (void)updateVisibility {
    CGRect rect = (CGRect){self.contentOffset, self.bounds.size};
    
    for (StreamItem *item in _items) {
        
        BOOL visible = CGRectIntersectsRect(item.frame, rect);
        if (item.visible != visible) {
            item.visible = visible;
            if (visible) {
                StreamReusableView *view = [self viewForItem:item];
                if (view) {
                    [self insertSubview:view atIndex:0];
                }
            } else {
                StreamReusableView *view = item.view;
                if (view) {
                    [view removeFromSuperview];
                    [item.metrics.reusableViews addObject:view];
                }
            }
        }
    }
}

// MARK: - User Actions

- (void)tap:(UITapGestureRecognizer *)recognizer {
    [self touchedAt:[recognizer locationInView:self]];
}

- (void)touchedAt:(CGPoint)point {
    StreamItem *item = [self visibleItemAtPoint:point];
    if (item) {
        if (item.selected) {
            item.selected = NO;
            self.selectedItem = nil;
        } else {
            item.selected = YES;
            self.selectedItem.selected = NO;
            self.selectedItem = item;
        }
        
        if ([self.delegate respondsToSelector:@selector(streamView:entryAt:)]) {
            [item.metrics select:item entry:[self.delegate streamView:self entryAt:item.index]];
        }
    }
}

- (StreamItem *)visibleItemAtPoint:(CGPoint)point {
    return [self itemPassingTest:^BOOL(StreamItem *item) {
        return CGRectContainsPoint(item.frame, point);
    }];
}

- (StreamItem *)itemPassingTest:(BOOL (^)(StreamItem *))test {
    if (test) {
        for (StreamItem *item in _items) {
            if (test(item)) {
                return item;
            }
        }
    }
    return nil;
}

- (void)scrollToItem:(StreamItem *)item animated:(BOOL)animated {
    [self scrollRectToVisible:item.frame animated:animated];
}

@end

File: meWrap/Views/StreamView/StreamView.swift
//
//  StreamView.swift
//  16wrap
//
//  Created by Sergey Maximenko on 9/3/15.
//  Copyright (c) 2015 Ravenpod. All rights reserved.
//

import Foundation
import UIKit

enum ScrollDirection {
    case Unknown, Up, Down
}

var StreamViewCommonLocksChanged: String = "StreamViewCommonLocksChanged"

protocol StreamViewDataSource {
    func numberOfSections() -> Int
    func numberOfItemsIn(section: Int) -> Int
    func metricsAt(position: StreamPosition) -> [StreamMetricsProtocol]
    func didLayoutItem(item: StreamItem)
    func entryBlockForItem(item: StreamItem) -> (StreamItem -> AnyObject?)?
    func didChangeContentSize(oldContentSize: CGSize)
    func didLayout()
    func headerMetricsIn(section: Int) -> [StreamMetricsProtocol]
    func footerMetricsIn(section: Int) -> [StreamMetricsProtocol]
}

extension StreamViewDataSource {
    func numberOfSections() -> Int { return 1 }
    func didLayoutItem(item: StreamItem) { }
    func entryBlockForItem(item: StreamItem) -> (StreamItem -> AnyObject?)? { return nil }
    func didChangeContentSize(oldContentSize: CGSize) { }
    func didLayout() { }
    func headerMetricsIn(section: Int) -> [StreamMetricsProtocol] { return [] }
    func footerMetricsIn(section: Int) -> [StreamMetricsProtocol] { return [] }
}

final class StreamView: UIScrollView {
    
    lazy var layout: StreamLayout = StreamLayout()
    
    private var reloadAfterUnlock = false
    
    var locks: Int = 0
    
    static var locks: Int = 0
    
    private var items = [StreamItem]()
    
    var dataSource: StreamViewDataSource?
    
    private weak var placeholderView: PlaceholderView?
    
    var placeholderViewBlock: (() -> PlaceholderView)?
    
    override var contentInset: UIEdgeInsets  {
        didSet {
            if oldValue != contentInset && items.count == 1 && layout.finalized {
                reload()
            }
        }
    }
    
    deinit {
        delegate = nil
        unsubscribeFromOffsetChange()
        NSNotificationCenter.defaultCenter().removeObserver(self, name:StreamViewCommonLocksChanged, object:nil)
    }
    
    private func setup() {
        subscribeOnOffsetChange()
        NSNotificationCenter.defaultCenter().addObserver(self, selector: #selector(StreamView.locksChanged), name: StreamViewCommonLocksChanged, object: nil)
    }
    
    private func unsubscribeFromOffsetChange() {
        layer.removeObserver(self, forKeyPath:"bounds")
    }
    
    private func subscribeOnOffsetChange() {
        layer.addObserver(self, forKeyPath: "bounds", options: .New, context: nil)
    }
    
    var didScrollUp: (() -> ())?
    var didScrollDown: (() -> ())?
    
    var trackScrollDirection = false
    
    var direction: ScrollDirection = .Unknown {
        didSet {
            if direction != oldValue {
                if direction == .Up {
                    didScrollUp?()
                } else {
                    didScrollDown?()
                }
            }
        }
    }
    
    override func observeValueForKeyPath(keyPath: String?, ofObject object: AnyObject?, change: [String : AnyObject]?, context: UnsafeMutablePointer<Void>) {
        if trackScrollDirection && tracking && (contentSize.height > height || direction == .Up) {
            direction = panGestureRecognizer.translationInView(self).y > 0 ? .Down : .Up
        }
        if layout.finalized {
            updateVisibility()
        }
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }
    
    private func clear() {
        placeholderView?.removeFromSuperview()
        for item in items {
            if let view = item.view {
                view.hidden = true
                item.metrics.enqueueView(view)
            }
        }
        items.removeAll()
    }
    
    class func lock() {
        locks += 1
    }
    
    class func unlock() {
        if (locks > 0) {
            locks -= 1
            NSNotificationCenter.defaultCenter().postNotificationName(StreamViewCommonLocksChanged, object: nil)
        }
    }
    
    func locksChanged() {
        if (locks == 0 && StreamView.locks == 0 && reloadAfterUnlock) {
            reloadAfterUnlock = false
            reload()
        }
    }
    
    func lock() {
        locks += 1
    }
    
    func unlock() {
        if (locks > 0) {
            locks -= 1
            locksChanged()
        }
    }
    
    func reload() {
        
        if locks > 0 || StreamView.locks > 0 {
            reloadAfterUnlock = true
            return
        }
        
        clear()
        
        layout.prepareLayout(self)
        
        addItems()
        
        if let item = items.last {
            changeContentSize(layout.contentSize(item, streamView: self))
        } else {
            if layout.horizontal {
                changeContentSize(CGSizeMake(0, height))
            } else {
                changeContentSize(CGSizeMake(width, 0))
            }
        }
        
        layout.finalizeLayout()
        
        _layoutSize = layoutSize(layer.bounds)
        
        dataSource?.didLayout()
        
        updateVisibility()
    }
    
    private func changeContentSize(newContentSize: CGSize) {
        let oldContentSize = contentSize
        if !CGSizeEqualToSize(newContentSize, oldContentSize) {
            contentSize = newContentSize
            dataSource?.didChangeContentSize(oldContentSize)
        }
    }
    
    private func addItems() {
        
        guard let dataSource = dataSource else { return }
        let layout = self.layout
        
        for section in 0..<dataSource.numberOfSections() {
            
            let position = StreamPosition(section: section, index: 0)
            for header in dataSource.headerMetricsIn(section) {
                addItem(metrics: header, position: position)
            }
            
            for i in 0..<dataSource.numberOfItemsIn(section) {
                let position = StreamPosition(section: section, index: i);
                for metrics in dataSource.metricsAt(position) {
                    if let item = addItem(dataSource, metrics: metrics, position: position) {
                        dataSource.didLayoutItem(item)
                    }
                }
            }
            
            for footer in dataSource.footerMetricsIn(section) {
                addItem(metrics: footer, position: position)
            }
            
            layout.prepareForNextSection()
        }
        
        if items.isEmpty, let placeholder = placeholderViewBlock {
            let placeholderView = placeholder()
            add(placeholderView, { (make) in
                make.centerX.equalTo(self)
                make.centerY.equalTo(self).offset(layout.offset/2)
                make.size.lessThanOrEqualTo(self).offset(-24)
            })
            self.placeholderView = placeholderView
        }
    }
    
    private func addItem(dataSource: StreamViewDataSource? = nil, metrics: StreamMetricsProtocol, position: StreamPosition) -> StreamItem? {
        let item = StreamItem(metrics: metrics, position: position)
        item.entryBlock = dataSource?.entryBlockForItem(item)
        metrics.modifyItem?(item)
        guard !item.hidden else { return nil }
        if let currentItem = items.last {
            item.previous = currentItem
            currentItem.next = item
        }
        layout.layoutItem(item, streamView: self)
        items.append(item)
        return item
    }
    
    private func updateVisibility() {
        updateVisibility(withRect: layer.bounds)
    }
    
    private var _layoutSize: CGFloat = 0
    
    private func layoutSize(rect: CGRect) -> CGFloat {
        return layout.horizontal ? rect.height : rect.width
    }
    
    private func reloadIfNeeded(rect: CGRect) -> Bool {
        let size = layoutSize(rect)
        if abs(_layoutSize - size) >= 1 {
            reload()
            return true
        } else {
            return false
        }
    }
    
    private func updateVisibility(withRect rect: CGRect) {
        guard !reloadIfNeeded(rect) else { return }
        for item in items {
            let visible = item.frame.intersects(rect)
            if item.visible != visible {
                item.visible = visible
                if visible {
                    let view = item.metrics.dequeueViewWithItem(item)
                    if view.superview != self {
                        insertSubview(view, atIndex: 0)
                    }
                    view.hidden = false
                } else if let view = item.view {
                    item.metrics.enqueueView(view)
                    view.hidden = true
                    item.view = nil
                }
            }
        }
    }
    
    // MARK: - User Actions
    
    func visibleItems() -> [StreamItem] {
        return itemsPassingTest { $0.visible }
    }
    
    func selectedItems() -> [StreamItem] {
        return itemsPassingTest { $0.selected }
    }
    
    var selectedItem: StreamItem? {
        return itemPassingTest { $0.selected }
    }
    
    func itemPassingTest(@noescape test: (StreamItem) -> Bool) -> StreamItem? {
        for item in items where test(item) {
            return item
        }
        return nil
    }
    
    func itemsPassingTest(@noescape test: (StreamItem) -> Bool) -> [StreamItem] {
        return items.filter(test)
    }
    
    func scrollToItemPassingTest(@noescape test: (StreamItem) -> Bool, animated: Bool) -> StreamItem? {
        let item = itemPassingTest(test)
        scrollToItem(item, animated: animated)
        return item
    }
    
    func scrollToItem(item: StreamItem?, animated: Bool)  {
        guard let item = item else { return }
        let minOffset = minimumContentOffset
        let maxOffset = maximumContentOffset
        if layout.horizontal {
            let offset = (item.frame.origin.x - contentInset.right) - fittingContentWidth / 2 + item.frame.size.width / 2
            if offset < minOffset.x {
                setContentOffset(minOffset, animated: animated)
            } else if offset > maxOffset.x {
                setContentOffset(maxOffset, animated: animated)
            } else {
                setContentOffset(CGPointMake(offset, 0), animated: animated)
            }
        } else {
            let offset = (item.frame.origin.y - contentInset.top) - fittingContentHeight / 2 + item.frame.size.height / 2
            if offset < minOffset.y {
                setContentOffset(minOffset, animated: animated)
            } else if offset > maxOffset.y {
                setContentOffset(maxOffset, animated: animated)
            } else {
                setContentOffset(CGPointMake(0, offset), animated: animated)
            }
        }
    }
    
    override func touchesShouldCancelInContentView(view: UIView) -> Bool {
        return true
    }
}